atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/expectations/ExpectationsManagerImpl.java

```diff
@@ -0,0 +1,68 @@
+/*
+ * (c) Copyright 2022 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.atlasdb.transaction.impl.expectations;
+
+import com.palantir.atlasdb.transaction.api.expectations.ExpectationsAwareTransaction;
+import com.palantir.logsafe.Preconditions;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+public final class ExpectationsManagerImpl implements ExpectationsManager {
+
+    private final Set<ExpectationsAwareTransaction> trackedTransactions = ConcurrentHashMap.newKeySet();
+    private final AtomicBoolean taskIsScheduled = new AtomicBoolean(false);
+    private final ScheduledExecutorService executorService;
+
+    public ExpectationsManagerImpl(ScheduledExecutorService executorService) {
+        this.executorService = executorService;
+    }
+
+    @Override
+    public void scheduleExpectationsTask(long delayMillis) {
+        Preconditions.checkArgument(
+                delayMillis > 0, "Transactional expectations manager scheduler delay must be strictly positive.");
+        if (!taskIsScheduled.compareAndExchange(false, true)) {
+            executorService.scheduleWithFixedDelay(
+                    new ExpectationsTask(trackedTransactions), delayMillis, delayMillis, TimeUnit.MILLISECONDS);
+        }
+    }
+
+    @Override
+    public void register(ExpectationsAwareTransaction transaction) {
+        trackedTransactions.add(transaction);
+    }
+
+    @Override
+    public void unregister(ExpectationsAwareTransaction transaction) {
+        trackedTransactions.remove(transaction);
+    }
+
+    @Override
+    public void markCompletion(ExpectationsAwareTransaction transaction) {
+        transaction.runExpectationsCallbacks();
+        unregister(transaction);
```

Comment:
So, do we need to make sure that the marking completion is idempotent? Or are there other protections in place for this to be called exactly once for a transaction lifecycle?
And how do we handle the case where callback throws an exception? We need to basically ensure that, unless the service itself dies, unregister happens.

Reply:
`register`/`unregister` are idempotent (as inherited from `Set<?>`)

For `markCompletion` to be idempotent, `runExpectationsCallback` should be idempotent too. I hadn't thought about that, I'll keep it in mind when I implement that part

For cases where tasks throw, the idea I had in prototype/drafts was to wrap the `runTaskWith*` methods in a try/finally and mark completion there. Same with `startTransactions` but use unregister if it fails